rust/lance-index/src/scalar/tag.rs

```diff
@@ -0,0 +1,289 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The Lance Authors
+
+use std::{
+    any::Any,
+    collections::{HashMap, HashSet},
+    fmt::Debug,
+    sync::Arc,
+};
+
+use arrow::array::AsArray;
+use arrow_array::{Array, RecordBatch, UInt64Array};
+use arrow_schema::{DataType, Field, Fields, Schema, SchemaRef};
+use async_trait::async_trait;
+use datafusion::physical_plan::{stream::RecordBatchStreamAdapter, SendableRecordBatchStream};
+use datafusion_common::ScalarValue;
+use deepsize::DeepSizeOf;
+use futures::TryStreamExt;
+use lance_core::{Error, Result};
+use roaring::RoaringBitmap;
+use snafu::{location, Location};
+
+use crate::{Index, IndexType};
+
+use super::{bitmap::train_bitmap_index, SargableQuery};
+use super::{
+    bitmap::BitmapIndex, btree::BtreeTrainingSource, AnyQuery, IndexStore, ScalarIndex, TagQuery,
+};
+
+pub const BITMAP_LOOKUP_NAME: &str = "bitmap_page_lookup.lance";
+
+trait TagSubIndex: ScalarIndex + DeepSizeOf {}
+
+impl<T: ScalarIndex + DeepSizeOf> TagSubIndex for T {}
+
+/// A scalar index that can be used on List<T> columns to
+/// support queries with array_contains_all and array_contains_any
+/// using an underlying bitmap index.
+#[derive(Clone, Debug, DeepSizeOf)]
+pub struct TagIndex {
+    values_index: Arc<dyn TagSubIndex>,
+}
+
+impl TagIndex {
+    fn new(values_index: Arc<dyn TagSubIndex>) -> Self {
+        Self { values_index }
+    }
+}
+
+#[async_trait]
+impl Index for TagIndex {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn as_index(self: Arc<Self>) -> Arc<dyn Index> {
+        self
+    }
+
+    fn as_vector_index(self: Arc<Self>) -> Result<Arc<dyn crate::vector::VectorIndex>> {
+        Err(Error::NotSupported {
+            source: "TagIndex is not a vector index".into(),
+            location: location!(),
+        })
+    }
+
+    fn index_type(&self) -> IndexType {
+        IndexType::Scalar
+    }
+
+    fn statistics(&self) -> Result<serde_json::Value> {
+        self.values_index.statistics()
+    }
+
+    async fn calculate_included_frags(&self) -> Result<RoaringBitmap> {
+        unimplemented!()
+    }
+}
+
+impl TagIndex {
+    async fn search_values(&self, values: &Vec<ScalarValue>) -> Result<Vec<UInt64Array>> {
+        let mut value_results = Vec::with_capacity(values.len());
+        for value in values {
+            let value_query = SargableQuery::Equals(value.clone());
+            value_results.push(self.values_index.search(&value_query).await?);
```

Comment:
tiny nit: could we return iterator here so we don't have to keep the results in memory can can instead union/intersect on the fly

Reply:
It's a bit trickier since the `values_index.search` method is async but I made it work with streams.